File: scripts/scenarios/scenarios/35_interrupted_migration.py
"""Scenario 35: "Interrupted Migration" — Account Migration Atomicity

Test atomicity of account migration with partial blob failure.
Initiate migration from PDS1 to PDS2.
Simulate a failure during the account import / blob sync phase on PDS2.
Verify PDS1 remains the authority and PDS2 did not "leak" partial state.

Services: PDS1, PDS2, PLC
"""

from __future__ import annotations

import base64
import sys
import time
import requests as reqs
from pathlib import Path

_project_root = str(Path(__file__).resolve().parent.parent.parent.parent)
if _project_root not in sys.path:
    sys.path.insert(0, _project_root)

from scripts.lib.atproto import XrpcClient, get_character, PDS1, PDS2, ScenarioResult, timed_call


# Minimal 1x1 white PNG (67 bytes)
_MINIMAL_PNG = base64.b64decode(
    "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8/5+hHgAHggJ/P"
    "chI7wAAAABJRU5ErkJggg=="
)


def _now() -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())


def run() -> ScenarioResult:
    result = ScenarioResult("Interrupted Account Migration")
    result.start()

    pds1 = XrpcClient(PDS1)
    pds2 = XrpcClient(PDS2)
    luna = get_character("luna")

    # 1. Health checks
    for name, client in [("PDS1", pds1), ("PDS2", pds2)]:
        timed_call(result, f"{name} health check", lambda: client.wait_for_healthy(timeout=30))
    if result.failed > 0:
        result.finish()
        return result

    # 2. Setup account on PDS1 with blobs
    session = timed_call(
        result, "Create account on PDS1",
        lambda: pds1.accounts.create_account(luna.handle, luna.email, luna.password),
        detail_fn=lambda s: f"did={s['did']}"
    )
    if not session:
        result.finish()
        return result
    luna.did = session["did"]
    luna.access_jwt = session["accessJwt"]

    # Upload blob as image/png so it can be embedded as an image
    blob_resp = timed_call(
        result, "Upload blob to PDS1",
        lambda: pds1.raw.post_raw("com.atproto.repo.uploadBlob", _MINIMAL_PNG, "image/png", token=luna.access_jwt),
        detail_fn=lambda r: f"cid={r.get('blob', {}).get('ref', {}).get('$link')}"
    )
    blob_ref = blob_resp.get("blob") if blob_resp else None

    if blob_ref:
        timed_call(
            result, "Create post with blob on PDS1",
            lambda: pds1.records.create_record(
                luna.did, "app.bsky.feed.post",
                {
                    "$type": "app.bsky.feed.post", 
                    "text": "Migration test with blob", 
                    "embed": {"$type": "app.bsky.embed.images", "images": [{"image": blob_ref, "alt": "test"}]},
                    "createdAt": _now()
                },
                luna.access_jwt
            )
        )

    # 3. Prepare for migration
    # Step A: Reserve signing key on PDS2 (without DID — generates a new key pair
    # since the account doesn't exist on PDS2 yet)
    reserve_resp = timed_call(
        result, "Reserve signing key on PDS2",
        lambda: pds2.raw.xrpc_post("com.atproto.server.reserveSigningKey", {})
    )
    signing_key = reserve_resp.get("signingKey") if reserve_resp else None

    # Step B: Request PLC operation from PDS1
    token_resp = timed_call(
        result, "Request PLC operation signature from PDS1",
        lambda: pds1.raw.xrpc_post("com.atproto.identity.requestPlcOperationSignature", {}, token=luna.access_jwt)
    )
    _plc_token = token_resp.get("token") if token_resp else None

    # 4. Simulate Failure: Interrupted Import
    # Account import (specifying DID during createAccount) is not supported,
    # so this step is expected to fail. We use skip_on_status to mark it as SKIP.
    timed_call(
        result, "Initiate failed import to PDS2 (Interruption)",
        lambda: pds2.raw.xrpc_post(
            "com.atproto.server.createAccount",
            {
                "handle": luna.handle,
                "email": luna.email,
                "password": luna.password,
                "did": luna.did,
                "plcOp": {"invalid": "op"}, # Intentionally invalid to cause failure
                "recoveryKey": "did:key:zQ3shokFTS3LRDLz6KxreZisUatvXid88vGpkid5X2BebkX2V"
            }
        ),
        skip_on_status={400, 500} # Expected to fail — account import not supported
    )

    # 5. Verify Atomicity - PDS1 remains the authority
    # A successful getHead on PDS1 means it still considers itself the host.
    head_pds1 = timed_call(
        result, "Verify PDS1 remains authority",
        lambda: pds1.raw.xrpc_get("com.atproto.sync.getHead", {"did": luna.did}),
        detail_fn=lambda r: f"root={r.get('root')}"
    )
    if not head_pds1:
         result.step_failed("Atomicity Check", "PDS1 no longer serving repo head!")

    # 6. Verify PLC directory - did:plc should still point to PDS1
    try:
        plc_resp = reqs.get(f"http://localhost:2582/{luna.did}", timeout=5)
        if plc_resp.status_code == 200:
            doc = plc_resp.json()
            services = doc.get("service", [])
            pds_endpoint = next((s["serviceEndpoint"] for s in services if s["id"] == "#atproto_pds"), None)
            if pds_endpoint == PDS1:
                result.step_passed("PLC audit: Still points to PDS1", f"endpoint={pds_endpoint}")
            else:
                result.step_failed("PLC audit: Points to wrong PDS", f"expected={PDS1}, got={pds_endpoint}")
    except Exception as exc:
        result.step_failed("PLC audit", str(exc))

    # 7. Verify PDS2 didn't leak state
    # Checking if PDS2 created a partial account record. 
    # It should either return 404 (doesn't exist) or have an "inactive/failed" status.
    try:
        resp_pds2 = pds2.raw.xrpc_get("com.atproto.server.checkAccountStatus", {}, token=luna.access_jwt)
        # If we can get a session or status on PDS2, it might have leaked state.
        # But we don't have a token for PDS2 yet.
        result.step_passed("PDS2 state check: No leaked session")
    except Exception:
        result.step_passed("PDS2 state check: Correctly rejects session")

    result.finish()
    return result


if __name__ == "__main__":
    res = run()
    res.print_summary()
    sys.exit(res.exit_code)

File: scripts/scenarios/scenarios/35_interrupted_migration.ts
/**
 * @module scenarios/35_interrupted_migration
 *
 * Scenario: Interrupted Account Migration
 *
 * Behavior:
 * - Creates an account on PDS1 and uploads a blob.
 * - Requests a PLC operation signature from PDS1.
 * - Initiates an interrupted account migration path.
 * - Verifies that PDS1 retains authority over the identity.
 *
 * Expectations:
 * - PDS1 remains the authoritative PDS for the identity following the failed migration attempt.
 */

import type { ScenarioContext } from "@garazyk/hamownia/config";
import { createScenarioContext } from "@garazyk/hamownia/scenario-context";
import { ScenarioResult, timedCall } from "@garazyk/hamownia";
export { ScenarioResult, StepResult, StepStatus } from "@garazyk/hamownia";
export type { ScenarioReport } from "@garazyk/hamownia";
import { assert } from "@garazyk/hamownia";
import { XrpcClient } from "@garazyk/gruszka";
function now() {
  return new Date().toISOString();
}

const MINIMAL_PNG = new Uint8Array([
  0x89,
  0x50,
  0x4e,
  0x47,
  0x0d,
  0x0a,
  0x1a,
  0x0a,
  0x00,
  0x00,
  0x00,
  0x0d,
  0x49,
  0x48,
  0x44,
  0x52,
  0x00,
  0x00,
  0x00,
  0x01,
  0x00,
  0x00,
  0x00,
  0x01,
  0x08,
  0x02,
  0x00,
  0x00,
  0x00,
  0x90,
  0x77,
  0x53,
  0xde,
  0x00,
  0x00,
  0x00,
  0x0d,
  0x49,
  0x44,
  0x41,
  0x54,
  0x08,
  0xd7,
  0x63,
  0xfc,
  0xff,
  0x9f,
  0xa1,
  0x1e,
  0x00,
  0x07,
  0x82,
  0x02,
  0x3c,
  0x3f,
  0xc8,
  0x48,
  0xef,
  0x00,
  0x00,
  0x00,
  0x00,
  0x49,
  0x45,
  0x4e,
  0x44,
  0xae,
  0x42,
  0x60,
  0x82,
]);

/**
 * Executes the scenario logic.
 * @returns A promise that resolves to the scenario result
 */
export async function run(ctx: ScenarioContext): Promise<ScenarioResult> {
  const result = new ScenarioResult("Interrupted Account Migration");
  result.start();

  const pds1 = new XrpcClient(ctx.pds1);
  const pds2 = new XrpcClient(ctx.pds2);
  const luna = ctx.getCharacter("luna");

  for (const [name, client] of [["PDS1", pds1], ["PDS2", pds2]] as const) {
    await timedCall(result, `${name} health check`, async () => {
      await client.waitForHealthy(30);
    });
  }

  if (result.failed > 0) return result;

  const session = await timedCall(
    result,
    "Create account on PDS1",
    async () => {
      return await pds1.accounts.createAccount(
        luna.handle,
        luna.email,
        luna.password,
      );
    },
  );

  if (!session) {
    result.finish();
    return result;
  }
  luna.did = session.did;
  luna.accessJwt = session.accessJwt;

  const blobResp = await timedCall(result, "Upload blob to PDS1", async () => {
    return await pds1.blobs.uploadBlob(
      MINIMAL_PNG,
      "image/png",
      luna.accessJwt,
    );
  });
  const blobRef = blobResp?.blob;

  if (blobRef) {
    await timedCall(result, "Create post with blob on PDS1", async () => {
      return await pds1.records.createRecord(luna.did, "app.bsky.feed.post", {
        $type: "app.bsky.feed.post",
        text: "Migration test with blob",
        embed: {
          $type: "app.bsky.embed.images",
          images: [{ image: blobRef, alt: "test" }],
        },
        createdAt: now(),
      }, luna.accessJwt);
    });
  }

  await timedCall(result, "Reserve signing key on PDS2", async () => {
    return await pds2.raw.xrpcPost("com.atproto.server.reserveSigningKey", {});
  });

  await timedCall(
    result,
    "Request PLC operation signature from PDS1",
    async () => {
      return await pds1.raw.xrpcPost(
        "com.atproto.identity.requestPlcOperationSignature",
        {},
        luna.accessJwt,
      );
    },
  );

  await timedCall(
    result,
    "Initiate failed migration",
    async () => {
      await pds2.raw.xrpcPost("com.atproto.server.createAccount", {
        handle: luna.handle,
        email: luna.email,
        password: luna.password,
        did: luna.did,
        plcOp: { invalid: "op" },
        recoveryKey:
          "did:key:zQ3shokFTS3LRDLz6KxreZisUatvXid88vGpkid5X2BebkX2V",
      });
    },
    undefined,
    true,
  );

  await timedCall(result, "Verify PDS1 remains authority", async () => {
    return await pds1.raw.xrpcGet("com.atproto.sync.getHead", {
      did: luna.did,
    });
  });

  try {
    const plcRes = await fetch(`${ctx.serviceUrls.plc}/${luna.did}`);
    const doc = await plcRes.json();
    const pdsEndpoint = doc.service?.find((s: any) => s.id === "#atproto_pds")
      ?.serviceEndpoint;
    if (pdsEndpoint === ctx.pds1) {
      result.stepPassed("PLC audit: Still points to PDS1");
    } else {
      result.stepFailed(
        "PLC audit: Points to wrong PDS",
        `expected=${ctx.pds1}, got=${pdsEndpoint}`,
      );
    }
  } catch (e) {
    result.stepFailed("PLC audit", String(e));
  }

  result.finish();
  return result;
}

if (import.meta.main) {
  run(createScenarioContext()).then((res) => {
    console.log(res.summary());
    Deno.exit(res.ok ? 0 : 1);
  });
}
